File: packages/mobile/android/app/src/main/java/com/pzero/BLEPeripheralModule.kt
package com.pzero

import android.Manifest
import android.bluetooth.*
import android.bluetooth.le.AdvertiseCallback
import android.bluetooth.le.AdvertiseData
import android.bluetooth.le.AdvertiseSettings
import android.bluetooth.le.BluetoothLeAdvertiser
import android.content.Context
import android.content.pm.PackageManager
import android.os.Build
import android.os.ParcelUuid
import android.util.Log
import androidx.core.app.ActivityCompat
import com.facebook.react.bridge.*
import org.json.JSONObject
import java.security.MessageDigest
import java.security.SecureRandom
import java.util.*

class BLEPeripheralModule(reactContext: ReactApplicationContext) :
    ReactContextBaseJavaModule(reactContext) {

    companion object {
        private const val TAG = "BLEPeripheralModule"
        private val SERVICE_UUID = UUID.fromString("550e8400-e29b-41d4-a716-446655440000")
        private val GET_ENDPOINTS_UUID = UUID.fromString("550e8400-e29b-41d4-a716-446655440001")
        private val GET_TOKEN_UUID = UUID.fromString("550e8400-e29b-41d4-a716-446655440002")
        private val DEVICE_PAIRING_UUID = UUID.fromString("550e8400-e29b-41d4-a716-446655440003")
    }

    private var bluetoothManager: BluetoothManager? = null
    private var bluetoothAdapter: BluetoothAdapter? = null
    private var bluetoothLeAdvertiser: BluetoothLeAdvertiser? = null
    private var bluetoothGattServer: BluetoothGattServer? = null

    private var isAdvertising = false
    private var startAdvertisingPromise: Promise? = null

    // Cached data
    private var endpointsData: ByteArray? = null
    private val tokensByEndpoint = mutableMapOf<String, ByteArray>()
    private val validEndpointIds = mutableSetOf<String>()
    private var requestedEndpointId: String? = null
    private var currentUid: String? = null

    // OOB pairing data
    private var oobRandomValue: ByteArray? = null
    private var oobTimestamp: Long? = null
    private val oobExpirySeconds: Long = 60

    override fun getName(): String = "BLEPeripheralModule"

    private fun checkBluetoothAdvertisePermission(): Boolean {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
            return ActivityCompat.checkSelfPermission(
                reactApplicationContext,
                Manifest.permission.BLUETOOTH_ADVERTISE
            ) == PackageManager.PERMISSION_GRANTED
        }
        return true
    }

    private fun checkBluetoothConnectPermission(): Boolean {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
            return ActivityCompat.checkSelfPermission(
                reactApplicationContext,
                Manifest.permission.BLUETOOTH_CONNECT
            ) == PackageManager.PERMISSION_GRANTED
        }
        return true
    }

    @ReactMethod
    fun initialize(promise: Promise) {
        try {
            bluetoothManager = reactApplicationContext.getSystemService(Context.BLUETOOTH_SERVICE) as BluetoothManager
            bluetoothAdapter = bluetoothManager?.adapter

            if (bluetoothAdapter == null) {
                promise.reject("BLUETOOTH_NOT_SUPPORTED", "Bluetooth not supported on this device")
                return
            }

            if (!bluetoothAdapter!!.isEnabled) {
                promise.reject("BLUETOOTH_DISABLED", "Bluetooth is disabled")
                return
            }

            bluetoothLeAdvertiser = bluetoothAdapter?.bluetoothLeAdvertiser

            if (bluetoothLeAdvertiser == null) {
                promise.reject("ADVERTISING_NOT_SUPPORTED", "BLE advertising not supported on this device")
                return
            }

            Log.d(TAG, "BLE Peripheral initialized successfully")
            promise.resolve(true)
        } catch (e: Exception) {
            Log.e(TAG, "Failed to initialize BLE Peripheral", e)
            promise.reject("INITIALIZATION_ERROR", e.message, e)
        }
    }

    @ReactMethod
    fun startAdvertising(endpointsJSON: String, promise: Promise) {
        try {
            if (bluetoothAdapter == null || bluetoothLeAdvertiser == null) {
                promise.reject("NOT_INITIALIZED", "BLE Peripheral not initialized")
                return
            }

            // Check permissions
            if (!checkBluetoothAdvertisePermission()) {
                promise.reject("PERMISSION_DENIED", "BLUETOOTH_ADVERTISE permission not granted")
                return
            }

            // Validate and store endpoints data
            try {
                val endpointsResponse = JSONObject(endpointsJSON)
                val endpointsArray = endpointsResponse.optJSONArray("data")

                // Extract valid endpoint IDs
                validEndpointIds.clear()
                if (endpointsArray != null) {
                    for (i in 0 until endpointsArray.length()) {
                        val endpoint = endpointsArray.getJSONObject(i)
                        val id = endpoint.optString("id")
                        if (id.isNotEmpty()) {
                            validEndpointIds.add(id)
                        }
                    }
                }

                Log.d(TAG, "Loaded ${validEndpointIds.size} valid endpoint IDs")
            } catch (e: Exception) {
                promise.reject("INVALID_JSON", "endpointsJSON is not valid JSON", e)
                return
            }
            endpointsData = endpointsJSON.toByteArray(Charsets.UTF_8)

            // Setup GATT server
            setupGattServer()

            // Start advertising
            val settings = AdvertiseSettings.Builder()
                .setAdvertiseMode(AdvertiseSettings.ADVERTISE_MODE_LOW_POWER)
                .setConnectable(true)
                .setTimeout(0)
                .setTxPowerLevel(AdvertiseSettings.ADVERTISE_TX_POWER_MEDIUM)
                .build()

            val data = AdvertiseData.Builder()
                .setIncludeDeviceName(true)
                .setIncludeTxPowerLevel(false)
                .addServiceUuid(ParcelUuid(SERVICE_UUID))
                .build()

            startAdvertisingPromise = promise
            bluetoothLeAdvertiser?.startAdvertising(settings, data, advertiseCallback)

            Log.d(TAG, "Initiated BLE advertising")
        } catch (e: Exception) {
            Log.e(TAG, "Failed to start advertising", e)
            promise.reject("ADVERTISING_ERROR", e.message, e)
        }
    }

    @ReactMethod
    fun stopAdvertising(promise: Promise) {
        try {
            if (bluetoothLeAdvertiser == null) {
                promise.resolve(true)
                return
            }

            // Check permissions
            if (!checkBluetoothAdvertisePermission()) {
                promise.reject("PERMISSION_DENIED", "BLUETOOTH_ADVERTISE permission not granted")
                return
            }

            bluetoothLeAdvertiser?.stopAdvertising(advertiseCallback)
            bluetoothGattServer?.close()
            bluetoothGattServer = null
            isAdvertising = false

            // Clear cached data
            tokensByEndpoint.clear()
            validEndpointIds.clear()
            requestedEndpointId = null
            currentUid = null

            // Clear OOB pairing data
            oobRandomValue = null
            oobTimestamp = null

            Log.d(TAG, "Stopped BLE advertising")
            promise.resolve(true)
        } catch (e: Exception) {
            Log.e(TAG, "Failed to stop advertising", e)
            promise.reject("STOP_ERROR", e.message, e)
        }
    }

    @ReactMethod
    fun isAdvertising(promise: Promise) {
        promise.resolve(isAdvertising)
    }

    @ReactMethod
    fun setTokenForEndpoint(endpointId: String, token: String, promise: Promise) {
        try {
            // Validate endpoint exists
            if (!validEndpointIds.contains(endpointId)) {
                promise.reject("INVALID_ENDPOINT", "Endpoint not found: $endpointId")
                return
            }

            val response = JSONObject()
            response.put("type", "token")
            response.put("data", token)

            tokensByEndpoint[endpointId] = response.toString().toByteArray(Charsets.UTF_8)

            Log.d(TAG, "Token set for endpoint: $endpointId")
            promise.resolve(true)
        } catch (e: Exception) {
            Log.e(TAG, "Failed to set token", e)
            promise.reject("TOKEN_ERROR", e.message, e)
        }
    }

    @ReactMethod
    fun transmitUid(uid: String, promise: Promise) {
        try {
            // Store the UID for transmission via BLE
            currentUid = uid

            val response = JSONObject()
            response.put("type", "device_pairing")
            response.put("uid", uid)
            response.put("timestamp", System.currentTimeMillis() / 1000)

            Log.d(TAG, "UID set for transmission: $uid")
            promise.resolve(true)
        } catch (e: Exception) {
            Log.e(TAG, "Failed to transmit UID", e)
            promise.reject("UID_ERROR", e.message, e)
        }
    }

    @ReactMethod
    fun generateOOBData(promise: Promise) {
        try {
            // Generate 128-bit (16 byte) cryptographically secure random value
            val randomBytes = ByteArray(16)
            SecureRandom().nextBytes(randomBytes)

            oobRandomValue = randomBytes
            oobTimestamp = System.currentTimeMillis()

            // Convert to hex string for transmission
            val randomHex = randomBytes.joinToString("") { "%02x".format(it) }

            // Generate confirmation value (SHA-256 of random value)
            val messageDigest = MessageDigest.getInstance("SHA-256")
            val confirmBytes = messageDigest.digest(randomBytes)
            val confirmHex = confirmBytes.joinToString("") { "%02x".format(it) }

            // Get device BLE address
            val deviceAddress = bluetoothAdapter?.address ?: UUID.randomUUID().toString()

            val oobData = Arguments.createMap().apply {
                putString("address", deviceAddress)
                putString("randomValue", randomHex)
                putString("confirmValue", confirmHex)
                putDouble("timestamp", (System.currentTimeMillis() / 1000).toDouble())
                putDouble("expirySeconds", oobExpirySeconds.toDouble())
            }

            Log.d(TAG, "Generated OOB pairing data")
            promise.resolve(oobData)
        } catch (e: Exception) {
            Log.e(TAG, "Failed to generate OOB data", e)
            promise.reject("CRYPTO_ERROR", e.message, e)
        }
    }

    @ReactMethod
    fun isOOBDataValid(promise: Promise) {
        val timestamp = oobTimestamp
        if (timestamp == null) {
            promise.resolve(false)
            return
        }

        val elapsed = (System.currentTimeMillis() - timestamp) / 1000
        promise.resolve(elapsed < oobExpirySeconds)
    }

    @ReactMethod
    fun clearOOBData(promise: Promise) {
        oobRandomValue = null
        oobTimestamp = null
        Log.d(TAG, "Cleared OOB pairing data")
        promise.resolve(true)
    }

    private fun setupGattServer() {
        if (!checkBluetoothConnectPermission()) {
            Log.e(TAG, "BLUETOOTH_CONNECT permission not granted")
            return
        }

        bluetoothGattServer = bluetoothManager?.openGattServer(reactApplicationContext, gattServerCallback)

        // Create characteristics with encryption required for security
        // This forces BLE pairing before any data can be read/written
        val getEndpointsCharacteristic = BluetoothGattCharacteristic(
            GET_ENDPOINTS_UUID,
            BluetoothGattCharacteristic.PROPERTY_READ or BluetoothGattCharacteristic.PROPERTY_WRITE,
            BluetoothGattCharacteristic.PERMISSION_READ_ENCRYPTED or BluetoothGattCharacteristic.PERMISSION_WRITE_ENCRYPTED
        )

        val getTokenCharacteristic = BluetoothGattCharacteristic(
            GET_TOKEN_UUID,
            BluetoothGattCharacteristic.PROPERTY_READ or BluetoothGattCharacteristic.PROPERTY_WRITE,
            BluetoothGattCharacteristic.PERMISSION_READ_ENCRYPTED or BluetoothGattCharacteristic.PERMISSION_WRITE_ENCRYPTED
        )

        val devicePairingCharacteristic = BluetoothGattCharacteristic(
            DEVICE_PAIRING_UUID,
            BluetoothGattCharacteristic.PROPERTY_READ or BluetoothGattCharacteristic.PROPERTY_WRITE or BluetoothGattCharacteristic.PROPERTY_NOTIFY,
            BluetoothGattCharacteristic.PERMISSION_READ_ENCRYPTED or BluetoothGattCharacteristic.PERMISSION_WRITE_ENCRYPTED
        )

        // Create service
        val service = BluetoothGattService(SERVICE_UUID, BluetoothGattService.SERVICE_TYPE_PRIMARY)
        service.addCharacteristic(getEndpointsCharacteristic)
        service.addCharacteristic(getTokenCharacteristic)
        service.addCharacteristic(devicePairingCharacteristic)

        // Add service to GATT server
        bluetoothGattServer?.addService(service)

        Log.d(TAG, "GATT server setup complete")
    }

    private val advertiseCallback = object : AdvertiseCallback() {
        override fun onStartSuccess(settingsInEffect: AdvertiseSettings?) {
            Log.d(TAG, "Advertising started successfully")
            isAdvertising = true
            startAdvertisingPromise?.resolve(true)
            startAdvertisingPromise = null
        }

        override fun onStartFailure(errorCode: Int) {
            Log.e(TAG, "Advertising failed with error code: $errorCode")
            isAdvertising = false
            startAdvertisingPromise?.reject("ADVERTISING_FAILED", "Failed with code: $errorCode")
            startAdvertisingPromise = null
        }
    }

    private val gattServerCallback = object : BluetoothGattServerCallback() {
        override fun onConnectionStateChange(device: BluetoothDevice?, status: Int, newState: Int) {
            if (newState == BluetoothProfile.STATE_CONNECTED) {
                Log.d(TAG, "Device connected: ${device?.address}")
            } else if (newState == BluetoothProfile.STATE_DISCONNECTED) {
                Log.d(TAG, "Device disconnected: ${device?.address}")
            }
        }

        override fun onCharacteristicReadRequest(
            device: BluetoothDevice?,
            requestId: Int,
            offset: Int,
            characteristic: BluetoothGattCharacteristic?
        ) {
            Log.d(TAG, "Read request for characteristic: ${characteristic?.uuid}")

            if (!checkBluetoothConnectPermission()) {
                bluetoothGattServer?.sendResponse(
                    device,
                    requestId,
                    BluetoothGatt.GATT_FAILURE,
                    offset,
                    null
                )
                return
            }

            var responseData: ByteArray? = null

            when (characteristic?.uuid) {
                GET_ENDPOINTS_UUID -> {
                    Log.d(TAG, "Client requesting endpoints data")
                    responseData = endpointsData
                }
                GET_TOKEN_UUID -> {
                    val endpointId = requestedEndpointId
                    if (endpointId != null) {
                        Log.d(TAG, "Client requesting token data for endpoint: $endpointId")
                        responseData = tokensByEndpoint[endpointId]
                    } else {
                        Log.d(TAG, "No endpoint ID specified in token request")
                    }
                }
                DEVICE_PAIRING_UUID -> {
                    val uid = currentUid
                    if (uid != null) {
                        Log.d(TAG, "Client requesting device pairing data with UID: $uid")
                        val response = JSONObject()
                        response.put("type", "device_pairing")
                        response.put("uid", uid)
                        response.put("timestamp", System.currentTimeMillis() / 1000)
                        responseData = response.toString().toByteArray(Charsets.UTF_8)
                    } else {
                        Log.d(TAG, "No UID available for device pairing")
                    }
                }
            }

            if (responseData != null) {
                if (offset >= responseData.size) {
                    bluetoothGattServer?.sendResponse(
                        device,
                        requestId,
                        BluetoothGatt.GATT_INVALID_OFFSET,
                        offset,
                        null
                    )
                    return
                }

                val value = responseData.copyOfRange(offset, responseData.size)
                bluetoothGattServer?.sendResponse(
                    device,
                    requestId,
                    BluetoothGatt.GATT_SUCCESS,
                    offset,
                    value
                )
                Log.d(TAG, "Read request successful")
            } else {
                bluetoothGattServer?.sendResponse(
                    device,
                    requestId,
                    BluetoothGatt.GATT_FAILURE,
                    offset,
                    null
                )
                Log.d(TAG, "No data available for characteristic")
            }
        }

        override fun onCharacteristicWriteRequest(
            device: BluetoothDevice?,
            requestId: Int,
            characteristic: BluetoothGattCharacteristic?,
            preparedWrite: Boolean,
            responseNeeded: Boolean,
            offset: Int,
            value: ByteArray?
        ) {
            Log.d(TAG, "Write request for characteristic: ${characteristic?.uuid}")

            if (!checkBluetoothConnectPermission()) {
                if (responseNeeded) {
                    bluetoothGattServer?.sendResponse(
                        device,
                        requestId,
                        BluetoothGatt.GATT_FAILURE,
                        offset,
                        null
                    )
                }
                return
            }

            // Parse write request to extract endpoint ID for token requests
            when (characteristic?.uuid) {
                GET_TOKEN_UUID -> {
                    try {
                        if (value != null) {
                            val requestJson = JSONObject(String(value, Charsets.UTF_8))
                            val requestType = requestJson.optString("type")

                            if (requestType == "getToken") {
                                requestedEndpointId = requestJson.optString("endpointId", null)
                                Log.d(TAG, "Token request for endpoint: $requestedEndpointId")
                            }
                        }
                    } catch (e: Exception) {
                        Log.e(TAG, "Failed to parse token request", e)
                    }
                }
                DEVICE_PAIRING_UUID -> {
                    try {
                        if (value != null) {
                            val requestJson = JSONObject(String(value, Charsets.UTF_8))
                            val requestType = requestJson.optString("type")

                            if (requestType == "devicePairing") {
                                Log.d(TAG, "Device pairing request received")
                            }
                        }
                    } catch (e: Exception) {
                        Log.e(TAG, "Failed to parse device pairing request", e)
                    }
                }
            }

            // Acknowledge write requests
            if (responseNeeded) {
                bluetoothGattServer?.sendResponse(
                    device,
                    requestId,
                    BluetoothGatt.GATT_SUCCESS,
                    offset,
                    null
                )
            }
            Log.d(TAG, "Write request acknowledged")
        }
    }
}

File: packages/mobile/ios/pzero/BLEPeripheralModule.swift
import Foundation
import CoreBluetooth
import React
import CryptoKit
import UIKit

@objc(BLEPeripheralModule)
class BLEPeripheralModule: NSObject, CBPeripheralManagerDelegate {

  private var peripheralManager: CBPeripheralManager?
  private var service: CBMutableService?
  private var getEndpointsCharacteristic: CBMutableCharacteristic?
  private var getTokenCharacteristic: CBMutableCharacteristic?
  private var devicePairingCharacteristic: CBMutableCharacteristic?

  // Cached data for characteristics
  private var endpointsData: Data?
  private var tokensByEndpoint: [String: Data] = [:]
  private var validEndpointIds: Set<String> = []
  private var requestedEndpointId: String?
  private var currentUid: String?

  // OOB pairing data
  private var oobRandomValue: Data?
  private var oobTimestamp: Date?
  private let oobExpirySeconds: TimeInterval = 60.0

  // BLE Service and Characteristic UUIDs (must match verifier)
  private let serviceUUID = CBUUID(string: "550e8400-e29b-41d4-a716-446655440000")
  private let getEndpointsUUID = CBUUID(string: "550e8400-e29b-41d4-a716-446655440001")
  private let getTokenUUID = CBUUID(string: "550e8400-e29b-41d4-a716-446655440002")
  private let devicePairingUUID = CBUUID(string: "550e8400-e29b-41d4-a716-446655440003")

  override init() {
    super.init()
  }

  // MARK: - React Native Module Setup

  @objc
  static func moduleName() -> String! {
    return "BLEPeripheralModule"
  }

  @objc
  static func requiresMainQueueSetup() -> Bool {
    return true
  }

  // MARK: - Public API

  @objc
  func initialize(_ resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) {
    DispatchQueue.main.async {
      self.peripheralManager = CBPeripheralManager(delegate: self, queue: nil)
      resolve(true)
    }
  }

  @objc
  func startAdvertising(_ endpointsJSON: String, resolver resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) {
    guard let manager = peripheralManager else {
      reject("NOT_INITIALIZED", "Peripheral manager not initialized", nil)
      return
    }

    guard manager.state == .poweredOn else {
      reject("BLUETOOTH_OFF", "Bluetooth is not powered on", nil)
      return
    }

    // Store endpoints data and extract valid endpoint IDs
    if let data = endpointsJSON.data(using: .utf8) {
      self.endpointsData = data

      // Parse and extract valid endpoint IDs
      do {
        if let json = try JSONSerialization.jsonObject(with: data, options: []) as? [String: Any],
           let endpointsArray = json["data"] as? [[String: Any]] {
          validEndpointIds.removeAll()
          for endpoint in endpointsArray {
            if let id = endpoint["id"] as? String, !id.isEmpty {
              validEndpointIds.insert(id)
            }
          }
          print("BLE Peripheral: Loaded \(validEndpointIds.count) valid endpoint IDs")
        }
      } catch {
        print("BLE Peripheral: Failed to parse endpoints JSON: \(error)")
      }
    }

    // Setup GATT service and characteristics
    setupService()

    // Start advertising
    let advertisementData: [String: Any] = [
      CBAdvertisementDataServiceUUIDsKey: [serviceUUID],
      CBAdvertisementDataLocalNameKey: "PZero Mobile"
    ]

    manager.startAdvertising(advertisementData)

    resolve(true)
  }

  @objc
  func stopAdvertising(_ resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) {
    guard let manager = peripheralManager else {
      reject("NOT_INITIALIZED", "Peripheral manager not initialized", nil)
      return
    }

    manager.stopAdvertising()

    if let service = service {
      manager.remove(service)
    }

    // Clear cached data to avoid stale reads
    self.endpointsData = nil
    self.tokensByEndpoint.removeAll()
    self.validEndpointIds.removeAll()
    self.requestedEndpointId = nil
    self.currentUid = nil

    // Clear OOB pairing data
    self.oobRandomValue = nil
    self.oobTimestamp = nil

    resolve(true)
  }

  @objc
  func isAdvertising(_ resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) {
    guard let manager = peripheralManager else {
      resolve(false)
      return
    }

    resolve(manager.isAdvertising)
  }

  @objc
  func transmitUid(_ uid: String, resolver resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) {
    // Store the UID for transmission via BLE
    self.currentUid = uid
    
    // Create response object
    let response: [String: Any] = [
      "type": "device_pairing",
      "uid": uid,
      "timestamp": Date().timeIntervalSince1970
    ]
    
    if let jsonData = try? JSONSerialization.data(withJSONObject: response, options: []) {
      print("BLE Peripheral: UID set for transmission: \(uid)")
      
      // If we have connected centrals, notify them of the new data
      if let characteristic = devicePairingCharacteristic,
         let manager = peripheralManager {
        manager.updateValue(jsonData, for: characteristic, onSubscribedCentrals: nil)
      }
      
      resolve(true)
    } else {
      reject("SERIALIZATION_ERROR", "Failed to serialize UID data", nil)
    }
  }

  @objc
  func generateOOBData(_ resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) {
    // Generate 128-bit (16 byte) cryptographically secure random value
    var randomBytes = [UInt8](repeating: 0, count: 16)
    let result = SecRandomCopyBytes(kSecRandomDefault, randomBytes.count, &randomBytes)

    guard result == errSecSuccess else {
      reject("CRYPTO_ERROR", "Failed to generate secure random bytes", nil)
      return
    }

    let randomData = Data(randomBytes)
    self.oobRandomValue = randomData
    self.oobTimestamp = Date()

    // Convert to hex string for transmission
    let randomHex = randomData.map { String(format: "%02x", $0) }.joined()

    // Generate a simple confirmation value (hash of random value)
    // In production, this would use proper BLE OOB confirmation calculation
    let confirmData = randomData.sha256()
    let confirmHex = confirmData.map { String(format: "%02x", $0) }.joined()

    // Get device identifier (we use a stable app-specific identifier since iOS doesn't expose BLE address)
    let deviceId = UIDevice.current.identifierForVendor?.uuidString ?? UUID().uuidString

    let oobData: [String: Any] = [
      "address": deviceId,
      "randomValue": randomHex,
      "confirmValue": confirmHex,
      "timestamp": Date().timeIntervalSince1970,
      "expirySeconds": oobExpirySeconds
    ]

    print("BLE Peripheral: Generated OOB pairing data")
    resolve(oobData)
  }

  @objc
  func isOOBDataValid(_ resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) {
    guard let timestamp = oobTimestamp else {
      resolve(false)
      return
    }

    let elapsed = Date().timeIntervalSince(timestamp)
    resolve(elapsed < oobExpirySeconds)
  }

  @objc
  func clearOOBData(_ resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) {
    oobRandomValue = nil
    oobTimestamp = nil
    print("BLE Peripheral: Cleared OOB pairing data")
    resolve(true)
  }

  @objc
  func setTokenForEndpoint(_ endpointId: String, token: String, resolver resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) {
    // Validate endpoint exists
    guard validEndpointIds.contains(endpointId) else {
      reject("INVALID_ENDPOINT", "Endpoint not found: \(endpointId)", nil)
      return
    }

    // Create response object
    let response: [String: Any] = [
      "type": "token",
      "data": token
    ]

    if let jsonData = try? JSONSerialization.data(withJSONObject: response, options: []) {
      self.tokensByEndpoint[endpointId] = jsonData
      print("BLE Peripheral: Token set for endpoint: \(endpointId)")
      resolve(true)
    } else {
      reject("SERIALIZATION_ERROR", "Failed to serialize token data", nil)
    }
  }

  // MARK: - Private Methods

  private func setupService() {
    // Create characteristics with encryption required for security
    // This forces BLE pairing before any data can be read/written
    getEndpointsCharacteristic = CBMutableCharacteristic(
      type: getEndpointsUUID,
      properties: [.read, .write],
      value: nil,
      permissions: [.readEncryptionRequired, .writeEncryptionRequired]
    )

    getTokenCharacteristic = CBMutableCharacteristic(
      type: getTokenUUID,
      properties: [.read, .write],
      value: nil,
      permissions: [.readEncryptionRequired, .writeEncryptionRequired]
    )

    devicePairingCharacteristic = CBMutableCharacteristic(
      type: devicePairingUUID,
      properties: [.read, .write, .notify],
      value: nil,
      permissions: [.readEncryptionRequired, .writeEncryptionRequired]
    )

    // Create service
    service = CBMutableService(type: serviceUUID, primary: true)

    // Safely unwrap characteristics
    guard let endpoints = getEndpointsCharacteristic, 
          let token = getTokenCharacteristic,
          let pairing = devicePairingCharacteristic else {
      print("BLE Peripheral: Failed to create characteristics")
      return
    }

    service?.characteristics = [endpoints, token, pairing]

    // Add service to peripheral manager
    guard let service = service else {
      print("BLE Peripheral: Failed to create service")
      return
    }
    peripheralManager?.add(service)
  }

  // MARK: - CBPeripheralManagerDelegate

  func peripheralManagerDidUpdateState(_ peripheral: CBPeripheralManager) {
    switch peripheral.state {
    case .poweredOn:
      print("BLE Peripheral: Powered On")
    case .poweredOff:
      print("BLE Peripheral: Powered Off")
    case .resetting:
      print("BLE Peripheral: Resetting")
    case .unauthorized:
      print("BLE Peripheral: Unauthorized")
    case .unsupported:
      print("BLE Peripheral: Unsupported")
    case .unknown:
      print("BLE Peripheral: Unknown")
    @unknown default:
      print("BLE Peripheral: Unknown state")
    }
  }

  func peripheralManager(_ peripheral: CBPeripheralManager, didAdd service: CBService, error: Error?) {
    if let error = error {
      print("BLE Peripheral: Error adding service: \(error.localizedDescription)")
      return
    }
    print("BLE Peripheral: Service added successfully")
  }

  func peripheralManagerDidStartAdvertising(_ peripheral: CBPeripheralManager, error: Error?) {
    if let error = error {
      print("BLE Peripheral: Error starting advertising: \(error.localizedDescription)")
      return
    }
    print("BLE Peripheral: Advertising started successfully")
  }

  func peripheralManager(_ peripheral: CBPeripheralManager, didReceiveRead request: CBATTRequest) {
    print("BLE Peripheral: Received read request for characteristic: \(request.characteristic.uuid)")

    var responseData: Data?

    if request.characteristic.uuid == getEndpointsUUID {
      responseData = endpointsData
    } else if request.characteristic.uuid == getTokenUUID {
      if let endpointId = requestedEndpointId {
        print("BLE Peripheral: Client requesting token data for endpoint: \(endpointId)")
        responseData = tokensByEndpoint[endpointId]
      } else {
        print("BLE Peripheral: No endpoint ID specified in token request")
      }
    } else if request.characteristic.uuid == devicePairingUUID {
      if let uid = currentUid {
        let response: [String: Any] = [
          "type": "device_pairing",
          "uid": uid,
          "timestamp": Date().timeIntervalSince1970
        ]
        responseData = try? JSONSerialization.data(withJSONObject: response, options: [])
        print("BLE Peripheral: Client requesting device pairing data with UID: \(uid)")
      } else {
        print("BLE Peripheral: No UID available for device pairing")
      }
    }

    if let data = responseData {
      if request.offset >= data.count {
        peripheral.respond(to: request, withResult: .invalidOffset)
        return
      }

      request.value = data.subdata(in: request.offset..<data.count)
      peripheral.respond(to: request, withResult: .success)
      print("BLE Peripheral: Read request successful")
    } else {
      peripheral.respond(to: request, withResult: .unlikelyError)
      print("BLE Peripheral: No data available for characteristic")
    }
  }

  func peripheralManager(_ peripheral: CBPeripheralManager, didReceiveWrite requests: [CBATTRequest]) {
    print("BLE Peripheral: Received write requests: \(requests.count)")

    for request in requests {
      guard let value = request.value else {
        peripheral.respond(to: request, withResult: .unlikelyError)
        continue
      }

      print("BLE Peripheral: Write request for \(request.characteristic.uuid)")

      // Parse the request JSON
      if let requestJSON = try? JSONSerialization.jsonObject(with: value, options: []) as? [String: Any],
         let requestType = requestJSON["type"] as? String {

        print("BLE Peripheral: Request type: \(requestType)")

        // Handle different request types
        if requestType == "getEndpoints" {
          // Endpoints data is already set when advertising started
          // The verifier will read it via didReceiveRead
          peripheral.respond(to: request, withResult: .success)
        } else if requestType == "getToken" {
          // Extract endpoint ID from request
          if request.characteristic.uuid == getTokenUUID {
            requestedEndpointId = requestJSON["endpointId"] as? String
            print("BLE Peripheral: Token request for endpoint: \(requestedEndpointId ?? "nil")")
          }
          // Token will be set by the app via setTokenForEndpoint
          // The verifier will read it via didReceiveRead
          peripheral.respond(to: request, withResult: .success)
        } else if requestType == "devicePairing" {
          // Handle device pairing request
          if request.characteristic.uuid == devicePairingUUID {
            print("BLE Peripheral: Device pairing request received")
            // The UID will be set by the app via transmitUid
            // The verifier will read it via didReceiveRead
          }
          peripheral.respond(to: request, withResult: .success)
        } else {
          peripheral.respond(to: request, withResult: .requestNotSupported)
        }
      } else {
        peripheral.respond(to: request, withResult: .unlikelyError)
      }
    }
  }

  func peripheralManager(_ peripheral: CBPeripheralManager, central: CBCentral, didSubscribeTo characteristic: CBCharacteristic) {
    print("BLE Peripheral: Central subscribed to characteristic: \(characteristic.uuid)")
  }

  func peripheralManager(_ peripheral: CBPeripheralManager, central: CBCentral, didUnsubscribeFrom characteristic: CBCharacteristic) {
    print("BLE Peripheral: Central unsubscribed from characteristic: \(characteristic.uuid)")
  }
}

// MARK: - Data Extension for SHA256

extension Data {
  func sha256() -> Data {
    let digest = SHA256.hash(data: self)
    return Data(digest)
  }
}
